ceno_zkvm/src/scheme/utils.rs

```diff
@@ -283,6 +297,212 @@ pub(crate) fn infer_tower_product_witness<E: ExtensionField>(
     wit_layers
 }
 
+pub fn build_main_witness<
```

Comment:
This flow share between cpu and gpu. The differentiation down to `layer_witness` level